### PR TITLE
Added code for fast fibonacci using Golden Ratio and Fast Exponentiation

### PR DESCRIPTION
**Pull request by vs666**
- [x] Read the [contribution guidelines][contrib-guidelines]
- [x] Added [unit tests][unit-tests] (or unit tests have already been added)
- [x] Updated the [Contents][contents]

[contrib-guidelines]: https://github.com/ProAlgos/ProAlgos-Cpp/blob/master/.github/CONTRIBUTING.md
[unit-tests]: https://github.com/ProAlgos/ProAlgos-Cpp/blob/master/.github/CONTRIBUTING.md#adding-unit-tests
[contents]: https://github.com/ProAlgos/ProAlgos-Cpp/blob/master/.github/README.md#contents

### CODE REVIEW

**Comment by alxmjo** [comment]
Thanks for your contribution, @vs666. Looks like the build is failing. Please fix that and then I'll move forward with the review.

**Comment by alxmjo** [comment]
And for what it's worth, I'd prefer a name besides `fibonacci_fast`. I'm not sure what a better name would be, but I think we should be specific about the strategy used (`fibonacci_golden`?) rather than just that it's "fast." Perhaps there's a clue somewhere in [here](https://en.wikipedia.org/wiki/Fibonacci_number#Binet's_formula).

**Comment by vs666** [comment]
@alxmjo I am unable to understand why my build is failing. I am relatively new, and could really use your advice.

**Comment by vs666** [comment]
@alxmjo build has passed, kindly review it.

**Comment by alxmjo** [comment]
Thank you. I'll have a look soon. In the meantime please update the [Contents](https://github.com/ProAlgos/ProAlgos-Cpp/blob/master/README.md#contents).

**Comment by vs666** [comment]
@alxmjo Contents have been updated. You can review the PR now.

**Comment by vs666** [comment]
@alxmjo any updates?

File: README.md [hunk 1]
```diff
@@ -62,6 +62,7 @@ handle it from there. :smile:
   * Nth Fibonacci number
     * [Linear time algorithm](cpp/include/algorithm/number_theory/fibonacci.hpp) :white_check_mark:
     * [Logarithmic time algorithm](cpp/include/algorithm/number_theory/fibonacci_efficient.hpp) :white_check_mark:
+    * [Golden Ratio (faster logarithmic time) algorithm](cpp/include/algorithm/number_theory/fibonacci_golden.hpp) :white_check_mark:
```
Comment (alxmjo) [inline README.md]: "Golden ratio algorithm". I don't think mentioning the timing is necessary.

File: cpp/CMakeLists.txt [hunk 2]
```diff
@@ -86,6 +87,11 @@ add_executable(fibonacci
         test/algorithm/number_theory/fibonacci.cpp)
 target_link_libraries(fibonacci test_runner)
 
+# Fast fibonacci using golden ratio
```
Comment (alxmjo) [inline cpp/CMakeLists.txt]: "Fast" is relative. I think just "Fibonacci using golden ratio" is better.

File: cpp/include/algorithm/number_theory/fibonacci_golden.hpp [hunk 3]
```diff
@@ -0,0 +1,73 @@
+/*
+    Fibonacci number
```
Comment (alxmjo) [inline cpp/include/algorithm/number_theory/fibonacci_golden.hpp]: For consistency, "Fibonacci using golden ratio"

File: cpp/include/algorithm/number_theory/fibonacci_golden.hpp [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    -----------------
```
Comment (alxmjo) [inline cpp/include/algorithm/number_theory/fibonacci_golden.hpp]: Make sure hyphens match text above.

File: cpp/include/algorithm/number_theory/fibonacci_golden.hpp [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+    Find the N-th Fibonacci number, given the value of N. The Fibonacci
+    sequence is characterized by the fact that every number after the
+    first two is the sum of the two preceding ones.
+
+    It uses the fact that the round off value of the next fibonacci number
+    is f_(n+1) = f_(n)*(phi) where phi is golden ratio
```
Comment (alxmjo) [inline cpp/include/algorithm/number_theory/fibonacci_golden.hpp]: Add period.

File: cpp/include/algorithm/number_theory/fibonacci_golden.hpp [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+        return n - 1;
+    }
+    else
+    {
+        return round(___modExp_____(n));
+    }
+
+}
+
+#endif // FIBONACCI_FAST_HPP
```
Comment (alxmjo) [inline cpp/include/algorithm/number_theory/fibonacci_golden.hpp]: Change to match name of file.

File: cpp/include/algorithm/number_theory/fibonacci_golden.hpp [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+#ifndef FIBONACCI_GOLDEN_HPP
+#define FIBONACCI_GOLDEN_HPP
+
+#include <cmath>
+typedef long long int ULL;
+
+const int MAX_N = 64; // fibonacci(64) cannot be calculated due to limitation of precision of computers for irrational number phi (GOLDEN RATIO)
+const long double phi = 1.618033988749894;
+
+long double ___modExp_____(const int ex) // Name is unconventional to prevent accidental clash with a function name in implementing class
```
Comment (alxmjo) [inline cpp/include/algorithm/number_theory/fibonacci_golden.hpp]: Can you expand on this? Why are the underscores necessary?

File: cpp/include/algorithm/number_theory/fibonacci_golden.hpp [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+{
+    int n=ex, exp=ex;
```
Comment (alxmjo) [inline cpp/include/algorithm/number_theory/fibonacci_golden.hpp]: Add spaces between identifiers.

File: cpp/include/algorithm/number_theory/fibonacci_golden.hpp [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+    while (exp)
+    {
+        if (exp & 1)
+        {
+            ans *= a;
+        }
+        a *= a;
+        exp >>= 1;
+    }
+    if (n%2==0) {
```
Comment (alxmjo) [inline cpp/include/algorithm/number_theory/fibonacci_golden.hpp]: `if (n % 2 == 0)`...

File: cpp/include/algorithm/number_theory/fibonacci_golden.hpp [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+        return sqrt(0.2)*(ans-1/(ans));
```
Comment (alxmjo) [inline cpp/include/algorithm/number_theory/fibonacci_golden.hpp]: `return sqrt(0.2) * (ans - 1 / (ans));`

Comment (alxmjo) [reply cpp/include/algorithm/number_theory/fibonacci_golden.hpp]: Fix other lines too, please.

File: cpp/test/algorithm/number_theory/fibonacci_golden.cpp [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+    REQUIRE(fibonacci(2) == 1);
+    REQUIRE(fibonacci(7) == 13);
+    REQUIRE(fibonacci(15) == 610);
+    REQUIRE(fibonacci(18) == 2584);
+    REQUIRE(fibonacci(23) == 28657);
+    REQUIRE(fibonacci(50) == 12586269025);
+    REQUIRE(fibonacci(64) == 10610209857723U);
+}
+
+TEST_CASE("Overflow cases", "[fibonacci]") {
```
Comment (alxmjo) [inline cpp/test/algorithm/number_theory/fibonacci_golden.cpp:19]: Thanks for adding these.

**Review by alxmjo — CHANGES_REQUESTED** [review]
Sorry it's taken me so long to get to this. Just a few minor requests regarding syntax and identifiers. Let me know if you have questions.

File: cpp/include/algorithm/number_theory/fibonacci_golden.hpp (hunk 7 above)
Comment (alxmjo) [inline cpp/include/algorithm/number_theory/fibonacci_golden.hpp]: Please add a comment explaining what this function does and how it works. Also, I'd see if you can come up with a more descriptive identifier than `modExp`.

File: cpp/include/algorithm/number_theory/fibonacci_golden.hpp [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+typedef long long int ULL;
+
+const int MAX_N = 64; // fibonacci(64) cannot be calculated due to limitation of precision of computers for irrational number phi (GOLDEN RATIO)
+const long double phi = 1.618033988749894;
+
+long double ___modExp_____(const int ex) // Name is unconventional to prevent accidental clash with a function name in implementing class
+{
+    int n=ex, exp=ex;
+    long double ans = 1, a = phi;
+    while (exp)
```
Comment (alxmjo) [inline cpp/include/algorithm/number_theory/fibonacci_golden.hpp:41]: In-line comments would be helpful here, too.

**Pull request closed without merging**